spec/truffle/interop/key_info_spec.rb

```diff
@@ -10,320 +10,524 @@
-      Truffle::Interop.key_info(@object, :ro).should include(:existing)
-      Truffle::Interop.key_info(@object, :rw).should include(:existing)
-      Truffle::Interop.key_info(@object, :wo).should include(:existing)
-    end
+          it "if the variable does not exist and the object is frozen" do
+            @object.freeze
+            Truffle::Interop.key_info(@object, :@foo).should_not include_any_of(:modifiable, :removable)
+          end
+      
+        end
+        
+      end
 
-    it "returns :readable for a readable name" do
-      Truffle::Interop.key_info(@object, :ro).should include(:readable)
-      Truffle::Interop.key_info(@object, :rw).should include(:readable)
-    end
+      describe "has INVOCABLE" do
+        
+        describe "not set" do
+          
+          it "if the variable exists and the object is not frozen" do
+            Truffle::Interop.key_info(@object, :@a).should_not include(:invocable)
+          end
+          
+          it "if the variable exists and the object is frozen" do
+            @object.freeze
+            Truffle::Interop.key_info(@object, :@a).should_not include(:invocable)
+          end
 
-    it "does not return :readable for a write-only name" do
-      Truffle::Interop.key_info(@object, :wo).should_not include(:readable)
-    end
+          it "if the variable does not exist and the object is not frozen" do
+            Truffle::Interop.key_info(@object, :@foo).should_not include(:invocable)
+          end
 
-    it "returns :writable for a writable name" do
-      Truffle::Interop.key_info(@object, :rw).should include(:writable)
-      Truffle::Interop.key_info(@object, :wo).should include(:writable)
-    end
+          it "if the variable does not exist and the object is frozen" do
+            @object.freeze
+            Truffle::Interop.key_info(@object, :@foo).should_not include(:invocable)
+          end
+      
+        end
+        
+      end
 
-    it "does not return :writable for a writable name for a frozen hash" do
-      @object.freeze
-      Truffle::Interop.key_info(@object, :rw).should_not include(:writable)
-      Truffle::Interop.key_info(@object, :wo).should_not include(:writable)
-    end
+      describe "has INTERNAL" do
+        
+        describe "set" do
+          
+          it "if the variable exists and the object is not frozen" do
+            Truffle::Interop.key_info(@object, :@a).should include(:internal)
+          end
+          
+          it "if the variable exists and the object is frozen" do
+            @object.freeze
+            Truffle::Interop.key_info(@object, :@a).should include(:internal)
+          end
 
-    it "does not return :writable for a read-only name" do
-      Truffle::Interop.key_info(@object, :ro).should_not include(:writable)
-    end
+          it "if the variable does not exist and the object is not frozen" do
+            Truffle::Interop.key_info(@object, :@foo).should include(:internal)
```

Comment:
This one seems odd to me. According to the javadocs for `KEY_INFO`: 

```
This message also allows a fast check of existence of a property among KEYS, the returned value is 0 iff the key does not exist.
```

Based on that reading, it seems to me `key_info` should return `:none` if the ivar doesn't exist.

Reply:
> the returned value is 0 iff the key does not exist

To me, this means that if the value is 0, then the key must not exist. It doesn't mean that if they key does not exist, then the value must be 0.

Also, if the value has to be 0 if the key doesn't exist, what would be the point of `INSERTABLE`?

> Single bit that is set if writing a new key is supported.